### ISSUE TITLE
all: support for declaring dependencies between issues

### ISSUE THREAD

**Issue by myitcv** [cue-lang/cue#1877]
As the CUE project grows in size, so will the number of issues (whether they be bug reports, feature requests or otherwise). Being able to declare dependencies of varying sorts between issues will make managing those issues easier.

Creating this issue as a WIP brain dump for discussions we have had regarding how to declare and surface dependencies between issues on GitHub. 

The types of dependencies we might want to declare include:

* `#n` is blocked on `#m`.
* `#n` is related to `#m`.
* `#n` is a duplicate of `#m`.
* ...

We asked for thoughts/suggestions in https://twitter.com/_myitcv/status/1562097305658052608.

Thus far this has generated the following suggestions (in addition to our own research):

* [snoozethis.com](https://twitter.com/Nevkontakte/status/1562111723502555137). However this does not surface/visualise the dependencies anywhere.
* [Creating task lists in an issue description](https://twitter.com/thp4/status/1562147280731475969). This is, however, manual, and only shows one end of the dependency in a consolidated place. 
* [Switching to GitLab](https://twitter.com/zygoon/status/1562132521890074624) 😄 
* [ZenHub](https://blog.zenhub.com/github-issue-dependencies/). Not yet explored. Idea originally from https://github.com/isaacs/github/issues/210#issuecomment-842660566.
* [CodeTree](https://codetree.com/). Not yet explored. Idea originally from https://github.com/isaacs/github/issues/210#issuecomment-842660566.
* Our own home-grown solution.

### Home-grown solution

Noting down a few rough ideas on how this might work.

* Use special "commands" in issue descriptions and comments to describe and update dependency graph.
* Create a GitHub App or similar that can listen to webhook events.
* On load, use the GitHub API to build up the dependency graph, a graph that is then augmented by webhook events. 
* Serve the dependency graph via a simple REST-like API in case we need it elsewhere.
* Update the description of issues that participate in the dependency graph by including an SVG image link, with a filename that is the hash of the current dependency graph for that issue. Render the dependency graph with that issue as the centre of the image. Update the image link (via GitHub API) each time the dependency graph for that issue changes.

**Comment by myitcv** [comment]
Updated the description to also include the fact we might want to use this mechanism to mark duplicates too, which might have default actions for closing the duplicate etc.

**Comment by mvdan** [comment]
GitHub does support marking issues as duplicates now, FYI: https://docs.github.com/en/issues/tracking-your-work-with-issues/marking-issues-or-pull-requests-as-a-duplicate

I used ZenHub as an external GitHub integration in a past job, via their browser extension which would load extra metadata to show on the sidebar. I would strongly recommend _against_ using something like that, because the experience was always rather slow and janky. Whereas vanilla GitHub generally loads one issue within 100-200ms for me, using ZenHub's extension meant multiple extra network roundtrips, so it would rather easily jump past a full second of page loading before all the information was visible. I also usually got the impression of feature creep, to the point that the sidebar had so much information that it wouldn't fit on my screen.

In terms of potential vanilla GitHub solutions - how about this for your three categories:

* "blocked by" uses a `blocked` label and has a comment near the end mentioning this.
* "related to" uses similar labels as the other related issues (e.g. `fmt` or `tooling`), and a comment near the end mentions this.
* "duplicate of" is closed with GitHub's native support, per the link above.

I think the main one here is "blocked by", since those issues should generally get priority. A label would allow us to find them quickly, prioritize them in a kanban-like board, etc.

**Comment by mvdan** [comment]
There are also relatively simple GitHub bot-like integrations that might be interesting to us: https://github.com/marketplace/actions/blocking-issues

**Comment by verdverm** [comment]
This could be contentious, but Jira is another option, seems like GitHub is trending in that direction, and there wouldn't be a need to migrate the code host. They have all the features and you can do some pretty cool things if you are a wizard.

I used ZenHub a few years back, it was pretty nice when I did, really nice in that it embeds itself in GitHub, more or less. GitHub has implemented some of their features since, enough that I haven't needed it for aggregation. I have no insights on issue dependencies for ZenHub.

**Comment by mvdan** [comment]
What I said about ZenHub being slow and its feature creep does apply to Jira to a much higher degree, in my limited experience :)

**Comment by mvdan** [comment]
Now that we have moved to GitHub Projects to plan projects, track milestones, and organize our upcoming work, we should also aim to use the upcoming native support for dependencies between issues via the (closed) tasklist beta: https://docs.github.com/en/issues/managing-your-tasks-with-tasklists/about-tasklists

That's not available and finished yet, it seems, but fingers crossed we can use it soon.

**Comment by mvdan** [comment]
GitHub added support for this via sub-issues, so that one issue can "hang" from another issue as a dependency or sub-task. I don't think we should use these everywhere, but at times they can be really useful.

https://docs.github.com/en/issues/tracking-your-work-with-issues/using-issues/adding-sub-issues
